### ISSUE TITLE
Implement multiple output file groups per processor

### ISSUE THREAD

**Issue by wrznr** [OCR-D/spec#117]
Pls. correct me if I am wrong: At the moment, a CLI invocation for an ocrd processor may only specify a single output file group. This is not sufficient for processors which may create multiple output data (i.e. XML and image files) in a single step (e.g. cropping, deskewing). In https://github.com/OCR-D/ocrd_tesserocr/pull/48 this is solved by setting a constant output file group name for the image files. This is a suboptimal solution. Most likely, the `-O` parameter has to be adjusted.

**Comment by bertsky** [comment]
I concur. This would not only be cleaner, but may be the only way to prevent conflicts (if multiple steps in the pipeline want to produce the same _kind_ of image data, but in a different way or from a different source). 

As with multiple input fileGrps, there is a slight complication in that usually, the position in the list might be meaningful and must be documented prominently. For example, one might have to specify `-O OCR-D-SEG-LINE-BIN,OCR-D-IMG-BIN` and _know_ that the second output will be used for the image data, just as in OCR evaluation one might have to specify `-I OCR-D-GT-SEG-LINE,OCR-D-OCR-OCRO,OCR-D-COR-ASV-ANN` and _know_ that the first position will be used as reference/GT.

**Comment by kba** [comment]
> At the moment, a CLI invocation for an ocrd processor may only specify a single output file group

The convention is the same for [--output-file-grp](https://ocr-d.github.io/cli#-o---output-file-grp-grp) as for input file groups, that multiple values can be specified, separated by comma.

Is there something preventing processor implementers from accepting multiple output file groups? Should core offer better support for parsing these comma-separated strings and always provide input/output file groups as an array?

I don't really see how to change the spec to accomodate this.

**Comment by wrznr** [comment]
I guess that if comma-separated values are already available, this issue can be closed and the processors may use them as @bertsky suggested.

**Comment by kba** [comment]
Actually, spec should be adapted to better document this in ocrd-tool.json.

@bertsky:

> Also, what do we do with output_file_grp in ocrd-tool.json? There needs to be a place where assumptions on input and output file groups can be put (how many, what order)... same for multi-OCR input or OCR-GT alignment/evaluation

Tricky because not only can there be multiple output file groups for different data types where order is important, but also different "variants" depending on the input.

Maybe change the structure of `output_file_grp` to a more expressive array of objects, e.g.:

```yaml
"ocrd-tesserocr-binarize": {
  "executable": "ocrd-tesserocr-binarize",
  "categories": ["Image preprocessing"],
  "description": "Binarize pages, regions or lines",
  "input_file_grp": [
    "OCR-D-IMG",
    "OCR-D-SEG-BLOCK",
    "OCR-D-SEG-LINE"
  ],
  "output_file_grp": [
    {"description": "PAGE-XML output", "examples": ["OCR-D-BIN-BLOCK", "OCR-D-BIN-LINE"], "required": true},
    {"description": "Binarized images", "default": "OCR-D-IMG-BIN"}
  ],
  "steps": ["preprocessing/optimization/binarization"],
  "parameters": {
    "operation_level": {
      "type": "string",
      "enum": ["region", "line"],
      "default": "region",
      "description": "PAGE XML hierarchy level to operate on"
    }
  }
}
```****

**Comment by bertsky** [comment]
Yes, something along those lines. But maybe this should be backwards compatible to the current format. (Is polymorphism allowed in JSON?) 

And I don't think we need both `description` _and_ `examples` – the latter could simply be part of the description. For many tools, this does not make much sense anyway: it fully depends on the overall configuration. For example, deskewing can be done...
- on the `OCR-D-IMG`, giving an `OCR-D-IMG-DESKEW` directly in METS `structMap`, or 
- on the `OCR-D-SEG-PAGE` or `OCR-D-GT-SEG-PAGE` (i.e. page file with `Border` from cropping), giving another `OCR-D-SEG-PAGE2` (or so) with `AlternativeImage` in `OCR-D-IMG-DESKEW`, or
- on the  `OCR-D-SEG-BLOCK` or `OCR-D-GT-SEG-BLOCK` (i.e. page file with regions), giving another `OCR-D-SEG-BLOCK2` (or so) with `AlternativeImage` in `OCR-D-IMG-DESKEW`, or
- on the  `OCR-D-SEG-LINE` or `OCR-D-GT-SEG-LINE` (i.e. page file with regions and lines), giving another `OCR-D-SEG-LINE2` (or so) with `AlternativeImage` in `OCR-D-IMG-DESKEW` etc.
  
...And these names could still be different if binarization and/or dewarping happened before. And sometimes it just depends on the requested `level-of-operation` / `operation_level` / `textequiv_level`.

It is exactly this kind of information that we need to make describable: what sensible configurations for the tool to run in (depending on the parameters) could be – in a free form, to encourage implementors to fully share their operational knowledge. Your `description` here looks like a good place for that.

Allowing a `default` is also a very good idea. (This behaviour would have to be validated, too.)

**Comment by kba** [comment]
#121 is closed, #59 implemented this. We've since moved away from different output file groups for PAGE-XML and (Alternative)images.
